src/librustc_metadata/creader.rs

```diff
@@ -608,7 +626,12 @@ impl<'a> CrateLoader<'a> {
             *(sym as *const &[ProcMacro])
         };
 
-        let extensions = decls.iter().map(|&decl| {
+        if macros.len() != decls.len() {
+            span_bug!(span, "proc-macro data mismatch: metadata '{:?}', decls {:?}'",
+                      macros.len(), decls.len());
+        }
+
+        let extensions = macros.zip(decls.iter()).map(|(proc_macro, &decl) | {
```

Comment:
I think `CrateMetadata` should hold onto `decls` (as a `&'static [ProcMacro]`) rather than the `FullProcMacro`s.